odl/deform/linearized_deform.py

```diff
@@ -0,0 +1,538 @@
+
+    """Adjoint operator of the derivative operator.
+
+    This operator computes the adjoint of the Gateaux derivative of the fixed
+    template linearized deformation operator for the fixed template ``I``
+    at given displacement ``v``:
+
+        J --> grad(I)(Id + v) J
+
+    where the ``J`` is the given template and
+    the ``Id`` is the identity mapping:
+
+        y --> y
+
+    and the vector field ``v`` is a displacement mapping:
+
+        y --> v(y)
+
+    Here, the ``y`` is an element in the domain of the target.
+    In diffeomorphic sense, the domain of ``I`` is the same
+    as the target.
+    """
+
+    def __init__(self, template, displacement):
+        """Initialize a new instance.
+
+        Parameters
+        ----------
+        template : `DiscreteLpVector`
+            Fixed template deformed by the vector field.
+        displacement: `ProductSpaceVector`
+            The point that the adjoint of the Gateaux derivative needs
+            to be computed at.
+
+        Examples
+        --------
+        >>> import odl
+        >>> space = odl.uniform_discr(0, 1, 5)
+        >>> template = space.element([0, 0, 1, 0, 0])
+        >>> disp_field_space = odl.ProductSpace(space, space.ndim)
+        >>> disp_field = disp_field_space.element([[0, 0, 0, -0.2, 0]])
+        >>> op = LinDeformFixedTemplDerivAdj(template, disp_field)
+        >>> given_template = space.element([1, 2, 1, 2, 1])
+        >>> op(given_template)
+        ProductSpace(uniform_discr(0.0, 1.0, 5), 1).element([
+            [0.0, 10.0, -5.0, -10.0, 0.0]
+        ])
+        """
+        if displacement[0] not in template.space:
+            raise TypeError('`displacement[0]` {!r} not an element of'
+                            '`template.space`'.format(displacement[0],
+                                                      template.space))
+
+        Operator.__init__(self, template.space, displacement.space,
+                          linear=True)
+
+        self.template = template
+        self.displacement = displacement
+
+    def _call(self, func):
+        """Implement ``self(func)```.
+
+        Parameters
+        ----------
+        func : `DiscreteLpVector`
+            The evaluation point, i.e. an element in the template space,
+            for the deformed gradient of the template.
+        """
+        grad = odl.Gradient(self.domain, method='forward',
+                            padding_method='symmetric')
+        template_grad = grad(self.template)
+
+        def_grad = deform_grad(template_grad, self.displacement)
+
+        return [gf * func for gf in def_grad]
+
+
+class LinDeformFixedDisp(Operator):
+
+    """Operator using fixed displacement to map template to deformed template.
+
+    This linear operator maps a template ``I`` to the corresponding
+    deformed template using a fixed displacement field ``v``.
+    The linearized deformation is given by the left action defined
+    through the pushforward, so
```

Comment:
How is the user supposed to know what "left action" and "pushforward" means? Either remove this part (recommended) or explain the notions (seems unnecessary).